### PR TITLE
sasgrid-and-viya Quickstart for Azure Resource Manager

### PR DESCRIPTION
**Pull request by cc-sshaik**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

*
*
*

### CODE REVIEW

**Comment by cc-sshaik** [comment]
Hi @bmoore-msft,

the pipeline got successful, could you please review and mark it for any changes?

**Comment by cc-sshaik** [comment]
@bmoore-msft could you please review this pull request and mark for any changes. 

Thanks For Advance.

File: sasgrid-and-viya/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,5833 @@
...
+                ]
+            }
+        },
+        {
+            "type": "Microsoft.KeyVault/vaults",
+            "name": "[variables('key_vault_name')]",
+            "apiVersion": "2019-09-01",
+            "location": "[parameters('location')]",
+            "properties": {
+                "enabledForDeployment": true,
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:636]: why is this needed?

File: sasgrid-and-viya/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,5833 @@
...
+                "enabledForDiskEncryption": false,
+                "enabledForTemplateDeployment": true,
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:638]: why is this needed?

File: sasgrid-and-viya/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('ansible_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('ansible_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:1911]: rbacPrincipalId should be the value of the principalId below

File: sasgrid-and-viya/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('meta_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('meta_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:2438]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('grid_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('grid_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:2745]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('mid_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('mid_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:3187]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', concat(variables('gridnode_vm_name'), copyIndex(1)))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', concat(variables('gridnode_vm_name'), copyIndex(1))), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:3541]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,5833 @@
...
+                        "name": "[concat(variables('oss_vm_name'),copyIndex(1),'-osdisk')]",
+                        "createOption": "FromImage",
+                        "caching": "ReadWrite",
+                        "managedDisk": {
+                            "storageAccountType": "Premium_LRS"
+                        },
+                        "diskSizeGB": "[variables('lustre_osdisk_size')]"
+                    },
+                    "dataDisks": [
+                        {
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:4382]: use a copy loop for these

File: sasgrid-and-viya/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('microservices_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('microservices_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:4788]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('spre_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('spre_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:5064]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,5833 @@
...
+                "[resourceId('Microsoft.Compute/virtualMachines', variables('Cascontroller_vm_name'))]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "templateLink": {
+                    "uri": "[variables('role_assignment_template')]"
+                },
+                "parameters": {
+                    "roleAssignmentName": {
+                        "value": "[guid(resourceId('Microsoft.Compute/virtualMachines', variables('Cascontroller_vm_name')), variables('reader_role'), variables('rbacPrincipalID'))]"
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:5346]: rbacPrincipalId should be the principalId value below

File: sasgrid-and-viya/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,5833 @@
...
+            "tags": "[variables('sasviyavm_tags')]",
+            "properties": {
+                "hardwareProfile": {
+                    "vmSize": "[parameters('viyaCASWorkerVmSize')]"
+                },
+                "storageProfile": {
+                    "imageReference": {
+                        "publisher": "RedHat",
+                        "offer": "RHEL",
+                        "sku
```
Comment (bmoore-msft) [inline sasgrid-and-viya/azuredeploy.json:5628]: rbacPrincipalId should be the principalId value below

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: sasgrid-and-viya/azuredeploy.json (hunk 8 above)
Comment (cc-sshaik) [reply sasgrid-and-viya/azuredeploy.json:4382]: We are provisioning only one data disk for  different virtual machines(mgt, mdt)  and hence didn't used copy loop.

File: sasgrid-and-viya/azuredeploy.json (hunk 1 above)
Comment (cc-sshaik) [reply sasgrid-and-viya/azuredeploy.json:636]: enabledForDeployment - to retrieve certificates from KeyValut we marked it as true. 
https://docs.microsoft.com/en-us/azure/templates/microsoft.keyvault/2018-02-14/vaults

File: sasgrid-and-viya/azuredeploy.json (hunk 2 above)
Comment (cc-sshaik) [reply sasgrid-and-viya/azuredeploy.json:638]: We have enabled these options to store our secrets and certificates in KeyValut and retrieve it from the template/scripts wherever it is required.

File: sasgrid-and-viya/azuredeploy.json (hunk 8 above)
Comment (cc-sshaik) [reply sasgrid-and-viya/azuredeploy.json:4382]: modified the code, used Copy loop.

**Comment by cc-sshaik** [comment]
Hi @bmoore-msft,

Could you please do a manual validation of this sample, and mark for any changes that need to be taken care of.  thanks!

File: sasgrid-and-viya/azuredeploy.json (hunk 3 above)
Comment (cc-sshaik) [reply sasgrid-and-viya/azuredeploy.json]: we have used rbacprincipalId, just as a variable for reference. Removed it to avoid confusion.

**Review by nahlaAlhebsi — APPROVED** [review]

**Review by nahlaAlhebsi — APPROVED** [review]

File: sasgrid-and-viya/azuredeploy.json (hunk 3 above)
Comment (bmoore-msft) [reply sasgrid-and-viya/azuredeploy.json]: Still not correct - the principalId that is used in the name must be the same as the principalId used in the assignment.  I.e. the managedIdentity value you have below

File: sasgrid-and-viya/azuredeploy.json (hunk 2 above)
Comment (bmoore-msft) [reply sasgrid-and-viya/azuredeploy.json:637]: Where are you doing template deployments that access the stored secrets?

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.
